engine/common/provider/engine.go

```diff
@@ -30,40 +39,81 @@ type RetrieveFunc func(flow.Identifier) (flow.Entity, error)
 // requests on the flow network. It is the `reply` part of the request-reply
 // pattern provided by the pair of generic exchange engines.
 type Engine struct {
-	unit     *engine.Unit
-	log      zerolog.Logger
-	metrics  module.EngineMetrics
-	me       module.Local
-	state    protocol.State
-	con      network.Conduit
-	channel  channels.Channel
-	selector flow.IdentityFilter
-	retrieve RetrieveFunc
+	component.Component
+	cm             *component.ComponentManager
+	log            zerolog.Logger
+	metrics        module.EngineMetrics
+	state          protocol.State
+	con            network.Conduit
+	channel        channels.Channel
+	requestHandler *engine.MessageHandler
+	requestQueue   mempool.EntityRequestStore
+	selector       flow.IdentityFilter
+	retrieve       RetrieveFunc
+	// buffered channel for EntityRequest workers to pick and process.
+	requestChannel chan *internal.EntityRequest
 }
 
+var _ network.MessageProcessor = (*Engine)(nil)
+
 // New creates a new provider engine, operating on the provided network channel, and accepting requests for entities
 // from a node within the set obtained by applying the provided selector filter. It uses the injected retrieve function
 // to manage the fullfilment of these requests.
-func New(log zerolog.Logger, metrics module.EngineMetrics, net network.Network, me module.Local, state protocol.State,
-	channel channels.Channel, selector flow.IdentityFilter, retrieve RetrieveFunc) (*Engine, error) {
+func New(
+	log zerolog.Logger,
+	metrics module.EngineMetrics,
+	net network.Network,
+	me module.Local,
+	state protocol.State,
+	requestQueue mempool.EntityRequestStore,
+	requestWorkers uint,
+	channel channels.Channel,
+	selector flow.IdentityFilter,
+	retrieve RetrieveFunc) (*Engine, error) {
 
-	// make sure we don't respond to requests sent by self or unauthorized nodes
+	// make sure we don't respond to request sent by self or unauthorized nodes
 	selector = filter.And(
 		selector,
 		filter.HasWeight(true),
 		filter.Not(filter.HasNodeID(me.NodeID())),
 	)
 
+	handler := engine.NewMessageHandler(
+		log,
+		engine.NewNotifier(),
+		engine.Pattern{
+			// Match is called on every new message coming to this engine.
+			// Provider engine only expects EntityRequest.
+			// Other message types are discarded by Match.
+			Match: func(message *engine.Message) bool {
+				request, ok := message.Payload.(*messages.EntityRequest)
+				if ok {
+					log.Info().
+						Str("entity_ids", fmt.Sprintf("%v", request.EntityIDs)).
+						Hex("requester_id", logging.ID(message.OriginID)).
```

Comment:
let's keep the field name consistent across messages for easier searching
```suggestion
						Hex("origin_id", logging.ID(message.OriginID)).
```